MGMT-130 competency models

dbhynds opened the pull request:
## Summary of Changes

* [x] **I acknowledge that this is a public repository, visible to anyone on the internet.**

<!--A few sentences describing the big picture of your changes. You can tag other engineers here who should see this PR even though they are not official reviewers.-->

lin-brian-l commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
 ## Engineer I
 
-**Technical Ability:** Engineer Is have a reasonable understanding of core engineering concepts. They have a basic understanding of the team's tools and processes and a broad introduction to engineering best practice and productivity skills. They have gained a high-level understanding of the architecture of the Packback codebase and design patterns, which they are able to demonstrate through thei ...
+An engineer II is capable of taking on well-defined technical problems and managing their own work each sprint.
```
Very minor typo here:

```suggestion
An engineer I is capable of taking on well-defined technical problems and managing their own work each sprint.
```

lin-brian-l commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Impact:** They take on work that helps the team meet its sprint goals. They actively participate in team norms, processes, and expectations. They focus on delivering the work assigned to them in a sprint.
 
-By the time an Engineer I is ready to be promoted to Engineer II, they will have expertise in Engineering best practice, tools, techniques and a solid introduction to the technologies in their domain.
+**Communication:** They ask questions to understand how their work fits into the larger context. They proactively communicate whether their work is ahead, on track, or behind and are especially mindful about communicating when work is taking longer than expected. They welcome feedback and apply the feedback they received in other contexts.
```
I think adding this extra comma after `behind` helps to clarify that the subsequent text is separate from the list.

```suggestion
**Communication:** They ask questions to understand how their work fits into the larger context. They proactively communicate whether their work is ahead, on track, or behind, and are especially mindful about communicating when work is taking longer than expected. They welcome feedback and apply the feedback they received in other contexts.
```

lin-brian-l commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Productivity:** They are capable of delivering a large volume of work regardless of the complexity/uncertainty of the work. They provide reliable estimates for features that take longer than a single sprint to complete.
 
-**Communication:** An Engineer III recognizes areas in our process that are redundant, manual, slow down work negatively, or allow for untested/unverified code to be pushed to live systems and offers suggestions to improve the process. An Engineer III knows that their work technically never "ends" in the sense that they always have the opportunity to recognize other opportunities for improvement  ...
+**Impact:** An Engineer III takes ownership over the "how" of the feature they implement with a "buck stops here" attitude. They are able to independently plan work for multiple engineers that will take several sprints to complete. They can identify strategic opportunities and risks with a proposed plan and affect the direction of the team. 
```
To be honest, I'm not sure what exactly you mean by a "'buck stops here' attitude"; is there another analogy you can use for clarity?

lin-brian-l commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
+The senior engineer displays a mastery of the core competencies of software engineering. Strong technical knowledge and high productivity are table stakes at this level. Growth opportunities focus on developing impact, communication, and collaboration skills in preparation for the staff engineer role.
 
-**Technical Ability:** The Senior Engineer should be seen as a solid engineer who is a go-to expert in at least one domain (e.g. frontend, backend, data, databases, site-reliability-engineering). The Senior Engineer is capable of owning technical design for projects of moderate complexity (measured in multiple sprints), and understands the trade-offs in creating good software in their area. They  ...
+**Technical Knowledge:** They have deep understanding of their technical specialty, and a working of knowledge of other domains. They take responsibility for the overall health and quality of their domain. They have responsibilities beyond writing code (managing audits, running on-call rotations, owning a backlog, etc.) to help them manage their domain within the organization. Feedback on their c ...
```
Minor addition to make the list clearer:

```suggestion
**Technical Knowledge:** They have deep understanding of their technical specialty, and a working of knowledge of other domains. They take responsibility for the overall health and quality of their domain. They have responsibilities beyond writing code (managing audits, running on-call rotations, owning a backlog, etc.) to help them manage their domain within the organization. Feedback on their code usually revolves around subjective decisions, rather than objective issues. They proactively stay abreast of industry best practices and trends, and have acquired at least one major skill outside of core coding such as monitoring, documentation, integration testing, visual design, and performance optimization.
```

lin-brian-l commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
+**Productivity:** A senior engineer regularly delivers a high volume of complex, critical work while also balancing other responsibilities or initiatives. The Senior Engineer requires very little oversight beyond high-level direction. They hold a depth of knowledge about the larger software ecosystem that enables them to debug quickly and effectively.
 
-The Senior Engineer has end-to-end responsibility for projects of increasing complexity that encompass more than their own development. They contribute to the common code bases and standards for the team. They understand the business that their code supports and use this knowledge to influence their task prioritization. Senior engineers assist in identifying and validating test cases and can iden ...
+**Impact:** Beyond just delivering work on our roadmap, a senior engineer helps implement innovations across our codebase, product, and department. They recognize and act upon opportunities that others miss. They seek to build objective and well-defended cases for decisions, and develop a reputation among the leadership team that their requests are objective, self-critical and have due-diligence  ...
```
I think this reads a little more clearly and gives a more active role on the engineer:

```suggestion
**Impact:** Beyond just delivering work on our roadmap, a senior engineer helps implement innovations across our codebase, product, and department. They recognize and act upon opportunities that others miss. They seek to build objective and well-defended cases for decisions, and develop a reputation among the leadership team that their requests are objective, self-critical, and have the due-diligence to understand all viable alternatives.
```

lin-brian-l commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
-**Technical Ability:** The Staff Engineer provides considerable high-level technical guidance across the team. They can usually anticipate and plan for technical problems. They are highly knowledgeable in major parts of our technology stack and are the technical owner of significant components of our code base. Staff engineers have a sustained track record of creating improvements in business-cri ...
+**Engineering Perspective:** Staff-plus engineers are the folks who will often get unexpectedly pulled into the room where this sort of decision is happening. This gives them the opportunity to inject the engineering context and perspective into a decision while it's still possible to change the outcome. These brief moments of input on critical decisions are unduly impactful and will allow you to ...
 
-**Technical Scope:** A Staff Engineer is still acting in a very hands-on role, and as such, they are a prolific contributor to both core projects at Packback and to experimental work. When presented with a complex problem, process or existing system they are able to reduce the complexity in order to get more done with less work. They provide guidance, direction and help colleagues build and produ ...
+Concrete examples could include:
 
-**Communication:** The Staff Engineer has strong abilities to influence without requiring reporting authority. They facilitate cross-team work and are influential beyond their individual group. They are capable of driving groups of disparate interests to decisions, and clearly communicating and seeing those decisions through to impact. The staff engineer is a very strong communicator and can simp ...
+- Your perspective is sought out in critical decisions
+- Actively inserting yourself into these critical moments
```
Shifting to the 2nd-person seems a little weird to me since every other section is in 3rd-person; can that be updated?

lin-brian-l commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
-**Communication:** The Staff Engineer has strong abilities to influence without requiring reporting authority. They facilitate cross-team work and are influential beyond their individual group. They are capable of driving groups of disparate interests to decisions, and clearly communicating and seeing those decisions through to impact. The staff engineer is a very strong communicator and can simp ...
+- Your perspective is sought out in critical decisions
+- Actively inserting yourself into these critical moments
+- Advocate for effective solutions and strategies
```
The other examples have active `-ing` verbs:

```suggestion
- Advocating for effective solutions and strategies
```

lin-brian-l commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
+
+A Staff Engineer is far more likely to change your company's long-term trajectory by growing other engineers than through personal heroics. Their effort here leads to a well-trained team with multiple experts in each domain.
+
+Effective mentorship is not a checkbox on the career ladder that can be satisfied by simply sharing advice and experience. Staff engineers build an ongoing relationship, understands the context of their mentee, and [sponsors](https://larahogan.me/blog/what-sponsorship-looks-like/) them.
```
Minor grammatical changes ("staff engineers" is plural):

```suggestion
Effective mentorship is not a checkbox on the career ladder that can be satisfied by simply sharing advice and experience. Staff engineers build an ongoing relationship, understand the context of their mentee, and [sponsor](https://larahogan.me/blog/what-sponsorship-looks-like/) them.
```

lin-brian-l commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
+Concrete examples could include:
+
+- Owning onboarding and continuing education initiatives
+- No bus-factor-one concerns / build redundancy in systems and knowledge
```
```suggestion
- No bus-factor-one concerns / building redundancy in systems and knowledge
```

lin-brian-l commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
+
+Concrete examples could include:
+
+- decreased onboarding time
+- got the team to agree on coding standards
+- reviewed designs and architecture decisions
+- made us build the right thing
+- made our customers happier
+- raised the organization's profile through public speaking/writing
+- simplified security and compliance concerns
```
The prior examples are all in the present tense, but these are in the past; how does this sound?

```suggestion
- decreasing onboarding time
- getting the team to agree on coding standards
- reviewing designs and architecture decisions
- making us build the right thing
- making our customers happier
- raising the organization's profile through public speaking/writing
- simplifying security and compliance concerns
```

lin-brian-l commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
+- raised the organization's profile through public speaking/writing
+- simplified security and compliance concerns
 
 ## Principal Engineer
```
I can't comment directly on the lines since they're outside the diff, but I've got some suggestions for this section.

1. 1st paragraph: add a comma after "QA":
> "(including but not limited to... engineering, QA, and data analysis)
2. 2nd paragraph: replace "is" with "are"
> "The problems that the Principal Engineer ~~is~~ are solving..."
3. 4th paragraph - this is very nit-picky, but I think "both" is a little confusing because one interpretation can be that there are two core projects, whereas it's actually referring to "Packback" and "side and experimental work". Maybe this is clearer?
> "...they are a prolific contributor to both Packback's core projects and their own external work."
4. 4th paragraph: I think adding some commas to this list would make it clearer:
> "When presented with a complex problem, process, or existing system, they..."
5. 4th paragrap: In the context of the words after this phrase, I think "leave" doesn't quite fit since it describes how the person should feel after the experience; IMO we should either swap it out with another word or tweak the subsequent phrase.

Swapping out "leave":
> "...working with this person should ~~leave~~ have team members feeling like..."

Tweaking the subsequent phrase:
> "...working with this person should leave team members feeling like they ~~are going to leave with~~ gained something significantly better than they came into.

6. 5th paragraph: I think repositioning the comma can make this sentence clearer.
> "They create architecture that shapes large parts of our business and ship complex projects, including many systems or major pieces of infrastructure."

7. 5th paragraph: The last sentence feels a little passive, what do you think of this?
> "They are the origin of multiple obvious and large technical contributions."

8. 6th paragraph: I think using semicolons instead of commas can make this sentence clearer.
> "... capable of setting short, medium, and long term strategic direction for parts of the technology stack; identifying areas of critical need based on future growth; and developing roadmaps to attack those problems."

lin-brian-l commented on careers/technical-engineer.md:
```diff
@@ -60,14 +90,10 @@ A Principal Engineer is still acting in a very hands-on role, and as such, they
-## Credits
-
-Packback's engineering ladder was inspired by documentation published by [Rent the Runway](https://dresscode.renttherunway.com/blog/ladder) and [Kickstarter](https://kickstarter.engineering/the-kickstarter-engineering-and-data-team-ladder-96996c3b327) on their engineering ladders. Thanks to them for open-sourcing that information.
+- Define strategic technical initiatives
+- Solve technical problems of the highest scope and complexity for the entire organization
+- Actively seek out and prioritize our toughest technical challenges with a goal of creating significant improvement
+- Define code quality standards and metrics
+- Provide mentorship to Senior and Staff Engineers to help them grow in their technical responsibilities
+- Ensure the safety and reliability of Packback's technology
+- Manage projects across multiple years
```
I'd personally prefer if these used active `-ing` verbs for consistency with other sections, but it's not a big deal either way.

lin-brian-l reviewed (commented):
@dbhynds great work, I've read through the docs and really like these changes! I have some suggestions for grammatical and wording changes, but the intent of the changes is sound.

codeunifier commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Technical Knowledge:** They have a robust understanding of their technical specialty, and a working knowledge of other domains. They can solve most problems in their domain. Their code is well-abstracted and performant as a rule and rarely needs a significant refactor at the PR stage. If an unexpected issue/bug does arise, they see developing a solution to that problem as their personal respons ...
 
-**Estimation:** An Engineer III Is able to provide reliable estimates for features that take longer than a single sprint to complete. Consistently delivers their work within a small margin of error to their original estimation, allowing the team to plan effectively and rely on the timely delivery of work.
+**Productivity:** They are capable of delivering a large volume of work regardless of the complexity/uncertainty of the work. They provide reliable estimates for features that take longer than a single sprint to complete.
```
I don't feel like this section is totally accurate. AFAIK we don't assign different capacities to engineers at different levels, so the "volume of work" wouldn't really change between levels, right? And we generally don't have tickets that take longer than a sprint to complete, so does `They provide reliable estimates for features that take longer than a single sprint to complete` refer to being a DRI?

dbhynds replied:
> Does `They provide reliable estimates for features that take longer than a single sprint to complete` refer to being a DRI?

Yes. Or any other technical planning

dbhynds replied:
> AFAIK we don't assign different capacities to engineers at different levels, so the "volume of work" wouldn't really change between levels, right?

Not on a sprint-by-sprint basis, no. That said, I would expect EN3s to be much more reliable with their work than EN2 or EN1. Aggregated over the course of a year, this does lead to a higher overall volume of work. Put another way... we give everyone (roughly) the same number of points. However, an EN1 is way more likely to get stuck, flounder, make questionable decisions, have to redo work after code review, etc. By EN3, these hiccups in productivity should mostly be ironed out, even with highly complex work. They should be able to reliably crank through a lot of work without difficulty.

codeunifier commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Productivity:** They are capable of delivering a large volume of work regardless of the complexity/uncertainty of the work. They provide reliable estimates for features that take longer than a single sprint to complete.
 
-**Communication:** An Engineer III recognizes areas in our process that are redundant, manual, slow down work negatively, or allow for untested/unverified code to be pushed to live systems and offers suggestions to improve the process. An Engineer III knows that their work technically never "ends" in the sense that they always have the opportunity to recognize other opportunities for improvement  ...
+**Impact:** An Engineer III takes ownership over the "how" of the feature they implement with a "buck stops here" attitude. They are able to independently plan work for multiple engineers that will take several sprints to complete. They can identify strategic opportunities and risks with a proposed plan and affect the direction of the team. 
```
`They are able to independently plan work for multiple engineers that will take several sprints to complete` - is this referring to being a DRI of a project as well? Or would this be similar to something like the accessibility initiative we've recently been doing?

I haven't reviewed your updates to the technical engineer file yet, but this whole section feels like something I'd expect of a Senior Engineer.

dbhynds replied:
DRI of a project could be one such scenario. I think coordinating accessibility or other such initiatives is another good example.

Generally speaking, I'd expect that an EN3 could plan the technical work for their team spanning a timeframe of a month or two. Whereas a senior engineer, I'd expect to be capable of effectively planning work spanning longer than 2 months.

codeunifier commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Communication:** They ask questions to understand how their work fits into the larger context. They proactively communicate whether their work is ahead, on track, or behind, and are especially mindful about communicating when work is taking longer than expected. They welcome feedback and apply the feedback they received in other contexts.
 
-## Engineer II
+**Collaboration:** They pair frequently. They are able to quickly recognize when they are stuck and ask for help. For more complicated tasks, they leverage the knowledge of their peers to plan and execute their work on time. They actively participate in code review and offer feedback to others.
```
Do Apprentices not get assigned PRs to review? It makes sense if that's the case, I just don't know if we have a mechanism in the auto-assigner to easily filter out developers at that level. 

FWIW, Aidan has been reviewing PRs and has done a pretty decent job so far.

dbhynds replied:
I do expect apprentices to participate in code review. However, due to their unconscious incompetencies, I wouldn't expect the reviews to be particularly insightful unless they pair with someone more senior.

In contrast, at EN1, I'd expect them to have the basic skills to offer constructive feedback -- or have the self-awareness to know where they need help reviewing.

I think your point about Aidan indicates that in this area they are performing beyond the level of apprentice.

codeunifier replied:
Maybe we could rephrase that last sentence to something like `They provide constructive feedback in code review and offer helpful feedback to others`?

dbhynds replied:
Modified it slightly to "They provide constructive feedback in code review and know when they need help constructing meaningful feedback."

codeunifier commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Communication:** They point out novel opportunities for refactoring during code review. They proactively work to minimize or mitigate potential problems before they arise.
 
-Packback's engineering ladder was inspired by documentation published by [Rent the Runway](https://dresscode.renttherunway.com/blog/ladder) and [Kickstarter](https://kickstarter.engineering/the-kickstarter-engineering-and-data-team-ladder-96996c3b327) on their engineering ladders. Thanks to them for open-sourcing that information.
+**Collaboration:** Others on the team reach out for help in their area of expertise, and they make themselves available. They are capable of effectively running small to medium-sized projects independently, but need coaching to successfully manage large projects.
```
`They are capable of effectively running small to medium-sized projects independently` - what does "running the project" mean in this context? I feel like other descriptions at Eng II suggest being a DRI at that level, so is this also in reference to being a DRI? 

I worry about the use of "independently" here because this is in the "Collaboration" section and there is a lot of collaboration with other team members when running any project.

dbhynds replied:
Yeah, again, leave the language open. DRI is one path, but it's not the only path.

Re: "independently" ... I guess what I'm trying to differentiate here is that they should have the skills to effectively run medium-sized projects effectively. If they're working on a large+ project, they'll probably need coaching from someone more senior to do it effectively.

Is there a better way to word that?

codeunifier replied:
I guess it just feels a little odd because in the context of a DRI, the Lead Reviewer is kind of meant to be someone more senior that could offer that coaching, right? So hopefully every DRI would have a coach that could help them, and "independently" feels like you're doing it on your own. Maybe we could reword it to something like:

`They are capable of effectively running small to medium-sized projects with little to no coaching, but need help to successfully manage large projects`?

codeunifier reviewed (changes requested):
I've just reviewed the Software Engineering stuff so far and wanted to get that stuff out for you before moving on to the next doc

codeunifier commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
+- Solve problems that span multiple groups
 
-**Technical Ability:** The Staff Engineer provides considerable high-level technical guidance across the team. They can usually anticipate and plan for technical problems. They are highly knowledgeable in major parts of our technology stack and are the technical owner of significant components of our code base. Staff engineers have a sustained track record of creating improvements in business-cri ...
+**Engineering Perspective:** Staff-plus engineers are the folks who will often get unexpectedly pulled into the room where this sort of decision is happening. This gives them the opportunity to inject the engineering context and perspective into a decision while it's still possible to change the outcome. These brief moments of input on critical decisions are unduly impactful and will allow you to ...
```
`this sort of decision` is unclear - what sort of decision?

codeunifier commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
+Concrete examples could include:
 
-**Communication:** The Staff Engineer has strong abilities to influence without requiring reporting authority. They facilitate cross-team work and are influential beyond their individual group. They are capable of driving groups of disparate interests to decisions, and clearly communicating and seeing those decisions through to impact. The staff engineer is a very strong communicator and can simp ...
+- Your perspective is sought out in critical decisions
```
Having an example of "Engineering Perspective" be "your perspective was given" doesn't feel like much of an example. Is there maybe an example of a critical decision where a Staff Engineer's perspective would be desired / required?

The other three examples here are kind of in the same boat - this feels more like restating the paragraph above rather than concrete examples.

codeunifier commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
+
+A Staff Engineer is far more likely to change your company's long-term trajectory by growing other engineers than through personal heroics. Their effort here leads to a well-trained team with multiple experts in each domain.
+
+Effective mentorship is not a checkbox on the career ladder that can be satisfied by simply sharing advice and experience. Staff engineers build an ongoing relationship, understand the context of their mentee, and [sponsor](https://larahogan.me/blog/what-sponsorship-looks-like/) them.
```
```suggestion
Effective mentorship is not a checkbox on the career ladder that can be satisfied by simply sharing advice and experience. Staff engineers build an ongoing relationship, understand the context of their mentee(s), and [sponsor](https://larahogan.me/blog/what-sponsorship-looks-like/) them.
```

codeunifier commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
+- Owning onboarding and continuing education initiatives
+- No bus-factor-one concerns / building redundancy in systems and knowledge
+- Providing critical feedback on architectural design documents
+- Sponsorship
```
Again, sponsorship is mentioned above. An example of sponsorship isn't sponsorship. Maybe we could say something like "Encouraging other engineers to take a specific online training or certification course"

dbhynds replied:
Lol. Good catch

codeunifier commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
-* Assume responsibility for fixing problems / "Always Responsible Individual"
-* Participate in Tier 2 on-call rotations
-* Managing projects across multiple trimesters
+Our teams are already optimized to incrementally improve. The role of the staff engineer is to be *transformative*. A staff engineer does more than improve their part of the codebase; they solve ambiguous, important problems that the company's current systems are ill-shaped to address. They tackle impossible technical problems (high ambiguity, no in-house expertise, poorly defined). They identify ...
```
I _love_ this explanation. Giving me goosebumps.

codeunifier reviewed (changes requested):
Got some feedback for the Technical Engineer doc now. Couple of very minor things and a few notes.

dbhynds commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Productivity:** They are capable of completing small, well-defined tasks on time with minimal help from other engineers. They are mindful of their technical limitations and take them into account when planning their work. They focus on reliably delivering quality work on time.
 
-**Communication:** They are learning how to communicate well and how to deliver feedback to peers and their manager. When given a task with unclear requirements they are learning how to ask for clarification and how to identify underlying assumptions in design and implementation and how to clarify them. They understand how their work fits into the larger picture for their team, and can work with  ...
+**Impact:** They take on work that helps the team meet its sprint goals. They actively participate in team norms, processes, and expectations. They focus on delivering the work assigned to them in a sprint.
```
Here are the differences that I'm trying to contrast by level:

- EN1: They focus on delivering the work assigned to them in a sprint.
- EN2: They are capable of independently planning work that multiple engineers can complete in a sprint or two.
- EN3: They are able to independently plan work for multiple engineers that will take several sprints to complete.

dbhynds replied:
I'm looking at the scale, timeline, and complexity of the work that they are responsible, increasing responsibility with seniority.

dbhynds commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Productivity:** They reliably deliver the work assigned to them in a sprint. When they encounter a complex problem, they leverage the knowledge of their peers to find a suitable solution, even if it means changing the scope of the work. They review their own code first before submitting a PR; they do not expect others to catch sloppy mistakes for them.
 
-**Communication:** They communicate well and are capable of delivering feedback to peers and their manager. When given a task with unclear requirements they know how to ask for clarification, and ensure that all assumptions are vetted before work starts to reduce the need for re-work. They understand how their work fits into the larger picture for their team, and use this to identify conflicting  ...
+**Impact:** They can take the lead on delivering sprint goals and are capable of independently planning work that multiple engineers can complete in a sprint or two. Instead of just executing on a task as written, they think critically about the problem and ensure that the solution will thoroughly solve the problem.
```
That's exactly what I was trying to say. I'll use your language here.

dbhynds commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
 # Technical Engineer Path
 
-The technical engineer focuses on individual contributorship and technical leadership
+The technical engineer focuses on individual contributorship and technical leadership. Their contributions level up the quality, effectiveness and efficiency of the entire organization. Whereas the [software engineer path](software-engineer.md) focused on mastering the fundamentals of software development using a defined set of criteria, the senior technical path focuses more on projects/efforts  ...
```
Love it.

dbhynds commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
+The senior engineer displays a mastery of the core competencies of software engineering. Strong technical knowledge and high productivity are table stakes at this level. Growth opportunities focus on developing impact, communication, and collaboration skills in preparation for the staff engineer role.
 
-**Technical Ability:** The Senior Engineer should be seen as a solid engineer who is a go-to expert in at least one domain (e.g. frontend, backend, data, databases, site-reliability-engineering). The Senior Engineer is capable of owning technical design for projects of moderate complexity (measured in multiple sprints), and understands the trade-offs in creating good software in their area. They  ...
+**Technical Knowledge:** They have deep understanding of their technical specialty, and a working of knowledge of other domains. They take responsibility for the overall health and quality of their domain. They have responsibilities beyond writing code (managing audits, running on-call rotations, owning a backlog, etc.) to help them manage their domain within the organization. Feedback on their c ...
```
EN3s have done it historically in practice. That said, usually it's our more senior EN3s who are trending towards promotion already

Jordanjc3 commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Productivity:** They are capable of completing small, well-defined tasks on time with minimal help from other engineers. They are mindful of their technical limitations and take them into account when planning their work. They focus on reliably delivering quality work on time. They review their own code first before submitting a PR.
 
-**Communication:** They are learning how to communicate well and how to deliver feedback to peers and their manager. When given a task with unclear requirements they are learning how to ask for clarification and how to identify underlying assumptions in design and implementation and how to clarify them. They understand how their work fits into the larger picture for their team, and can work with  ...
+**Impact:** They take on work that helps the team meet its sprint goals. They actively participate in team norms, processes, and expectations. They focus on delivering the work assigned to them in a sprint.
```
Similarly, I'd re-write this to reflect the outcomes, rather than the inputs. Anyone can "focus on delivering the work assigned to them", but the quality of the follow-through is part of what defines the different levels.

> They deliver well-defined tasks that contribute to the team's sprint goals. [second sentence stays the same]. They deliver the work assigned to them in sprint with [choose your modifier] consistency.

Jordanjc3 commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Technical Knowledge:** They have practical experience implementing the core concepts and best practices of software engineering. They are able to write performant, well-abstracted code with minimal help from other engineers. They are able to solve most problems in their domain, but may need help with complex tasks and tasks outside their domain.
 
-**Estimation:** Accountable and reliable to all commitments which they have taken ownership over, both in-office (by completing all work in the Sprint Backlog that they pulled in during Sprint Planning) and when On-Call (by promptly responding to incoming requests)
+**Productivity:** They reliably deliver the work assigned to them in a sprint. When they encounter a complex problem, they leverage the knowledge of their peers to find a suitable solution, even if it means changing the scope of the work. By the PR stage, their code is well-constructed; they do not expect others to catch sloppy mistakes for them.
```
I would change the last sentence here "they do not expect others to catch sloppy mistakes for them". What if they don't expect it, but others are frequently catching sloppy mistakes in their PRs? I'd rather this just be oriented towards their PRs being largely "sloppy mistake free".

Jordanjc3 commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Productivity:** They reliably deliver the work assigned to them in a sprint. When they encounter a complex problem, they leverage the knowledge of their peers to find a suitable solution, even if it means changing the scope of the work. By the PR stage, their code is well-constructed; they do not expect others to catch sloppy mistakes for them.
 
-**Communication:** They communicate well and are capable of delivering feedback to peers and their manager. When given a task with unclear requirements they know how to ask for clarification, and ensure that all assumptions are vetted before work starts to reduce the need for re-work. They understand how their work fits into the larger picture for their team, and use this to identify conflicting  ...
+**Impact:** They can take the lead on delivering work critical to meeting the sprint goal and are capable of independently planning work that multiple engineers can complete in a sprint or two. Instead of just executing on a task as written, they think critically about the problem and ensure that the solution will thoroughly solve the problem.
```
Are we suggesting that E II's are not capable of DRI'ing projects? Setting the capability at multiple engineers and a couple sprints of work falls short of what we'd require of a DRI.

dbhynds replied:
E2s can DRI projects! We have a handful for roadmap projects that are small-to-medium in size that can be accomplished on this timeline. I'd trust an E2 with a small project, and an E2 trending towards promotion with a medium project. I wouldn't trust them with a large unless there was a lot of hands-on involvement from a senior.

Jordanjc3 commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Productivity:** They reliably deliver the work assigned to them in a sprint. When they encounter a complex problem, they leverage the knowledge of their peers to find a suitable solution, even if it means changing the scope of the work. By the PR stage, their code is well-constructed; they do not expect others to catch sloppy mistakes for them.
 
-**Communication:** They communicate well and are capable of delivering feedback to peers and their manager. When given a task with unclear requirements they know how to ask for clarification, and ensure that all assumptions are vetted before work starts to reduce the need for re-work. They understand how their work fits into the larger picture for their team, and use this to identify conflicting  ...
+**Impact:** They can take the lead on delivering work critical to meeting the sprint goal and are capable of independently planning work that multiple engineers can complete in a sprint or two. Instead of just executing on a task as written, they think critically about the problem and ensure that the solution will thoroughly solve the problem.
```
Instead of "they can take the lead", can we describe how often they might take the lead? "They can" means they have the capability, but I'd expect "Impact" to articulate how they wield it.

Jordanjc3 replied:
For the last sentence, I'd try to make the first part a bit more visible. Instead of stating that they think about the task, how does that actually manifest itself? Do they work with Product/team lead/ DRI to make adjustments? Spark conversations with the team?

Jordanjc3 commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Impact:** They can take the lead on delivering work critical to meeting the sprint goal and are capable of independently planning work that multiple engineers can complete in a sprint or two. Instead of just executing on a task as written, they think critically about the problem and ensure that the solution will thoroughly solve the problem.
 
-Engineer IIs can take a bug or a task, and instead of just executing on the task as written, think critically about the problem and ensure that the issue has been thoroughly resolved by identifying root causes, proactively addressing performance concerns, providing proactive feedback on any product requirements and writing solid tests.
+**Communication:** They are able to effectively review code in their area of expertise and regularly point out opportunities for improvement in the code that they review. They proactively point out concerns with proposed or current sprint work as soon as they see them. They can articulate the tradeoffs of different approaches to solving a problem.
```
Can we re-write this first sentence to avoid "they are able to" and focus more on "they do" ?

Jordanjc3 commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Communication:** They are able to effectively review code in their area of expertise and regularly point out opportunities for improvement in the code that they review. They proactively point out concerns with proposed or current sprint work as soon as they see them. They can articulate the tradeoffs of different approaches to solving a problem.
 
-By the time an engineer is ready to be promoted to Engineer III they will have focused on a technology or functional area as their expertise and become capable of mentoring interns and new engineers in these areas.
+**Collaboration:** They are mindful of the work that others are doing and can critique or suggest improvements to the proposed plans of others. They get help from others when stuck, but are also capable of helping unstick their peers. They are capable of effectively running small to medium-sized projects with regular coaching.
```
Can we swap "they are mindful of" with something that describes visible behavior?

Jordanjc3 replied:
"can critique or suggest improvements" reflects ability, rather than behavior. Do Engineers at this level engage in that behavior? What do our expectations look like?

dbhynds replied:
The behavior I'm trying to describe here is:

Going beyond just the work assigned to them, they are aware of what the rest of the team is working on. When they see an opportunity for improvement in someone else's work, they speak up.

Say they're working on a piece of backend work and hit a piece of complexity. They are aware of how the FE is planning to consume this data. They realize with a slight modification to their work, they could significantly simplify the FE implementation for another team member.

Jordanjc3 commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
 ## Engineer III
 
-**Technical Ability:** An Engineer III can consistently be relied on to take on an undefined project through to execution while fully and effectively utilizing the skills and intelligence of the team. An Engineer III has the ability to provide an answer to most technical questions that the team may have, but they make a choice to exercise restraint and ask questions to help other team members rea ...
+An engineer III is capable of taking on complex technical problems and coordinating work spanning several months.
```
The usage of time-scales is probably OK here. But, what we really care about is which Engineers are capable of leading projects. Yet, saying "DRIs projects" is probably a bit harder to make clear since they came in all sizes and shapes.

Based on how this is written, I assume E III's are the when folks can start DRI'ing things (for the most part).

dbhynds replied:
We don't have hard and fast rules around this, but generally speaking, yes. E3+ is when we typically have people start DRIing large projects. And most of our projects are large.

Jordanjc3 commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+An engineer III is capable of taking on complex technical problems and coordinating work spanning several months.
 
-**Technical Scope:** An Engineer III takes ownership over the "how" of the feature they implement with a "buck stops here" attitude. If an issue/bug does arise, they see developing a solution to that problem as their personal responsibility. An Engineer III actively seeks out knowledge about the risks and challenges in our business, using that knowledge to guide decision-making
+**Technical Knowledge:** They have a robust understanding of their technical specialty, and a working knowledge of other domains. They can solve most problems in their domain. Their code is well-abstracted and performant as a rule and rarely needs a significant refactor at the PR stage. If an unexpected issue/bug does arise, they see developing a solution to that problem as their personal respons ...
```
I think you can just drop "as a rule" since you follow it up with noting it rarely needs a significant refactor (for any reason, abstraction or performance aside).

Jordanjc3 commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Productivity:** They are capable of delivering a large volume of work regardless of the complexity/uncertainty of the work. They provide reliable estimates for features that take longer than a single sprint to complete.
 
-**Communication:** An Engineer III recognizes areas in our process that are redundant, manual, slow down work negatively, or allow for untested/unverified code to be pushed to live systems and offers suggestions to improve the process. An Engineer III knows that their work technically never "ends" in the sense that they always have the opportunity to recognize other opportunities for improvement  ...
+**Impact:** An Engineer III takes ownership over the "how" of the feature they implement with a "buck stops here" attitude. They are able to independently plan work for multiple engineers that will take several sprints to complete. They can identify strategic opportunities and risks with a proposed plan and affect the direction of the team. 
```
I would write-out the "are able-to" and "they can" so we can re-write this to be more active.

Jordanjc3 commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Impact:** An Engineer III takes ownership over the "how" of the feature they implement with a "buck stops here" attitude. They are able to independently plan work for multiple engineers that will take several sprints to complete. They can identify strategic opportunities and risks with a proposed plan and affect the direction of the team. 
 
-## Credits
+**Communication:** They point out novel opportunities for refactoring during code review. They proactively work to minimize or mitigate potential problems before they arise.
```
For the second sentence, what does this look like in terms of communication? You could do the second bit completely independently (e.g. Matt finds pieces of risk and opens random PRs for them all the time), what are we looking to articulate here?

Jordanjc3 commented on careers/technical-engineer.md:
```diff
@@ -1,48 +1,78 @@
 # Technical Engineer Path
```
I would suggest that Senior+ Engineers certainly do take on sprint goal related work. They're not always DRI'ing or leading a project, it's possible that they're participating in a project that someone else is DRI'ing and as a result, are contributing to sprint goals.

This is how things currently work on my team.

Jordanjc3 commented on careers/technical-engineer.md:
```diff
@@ -1,73 +1,99 @@
 ## Senior Engineer
 
-Success as a Senior Engineer is based on their ability to make the team and individuals on the team more effective. A Senior Engineer's technical impact should be focused on using technology to increase the likelihood that Packback succeeds as a business. A Senior Engineer at Packback should be focused on making a 10x impact on the team; which can only be done by empowering others, optimizing pro ...
+The senior engineer displays a mastery of the core competencies of software engineering. Strong technical knowledge and high productivity are table stakes at this level. Growth opportunities focus on developing impact, communication, and collaboration skills in preparation for the staff engineer role.
```
I can't think of a better way to phrase this, so we should probably keep as is. Yet, having the growth opportunities oriented around getting promoted to the next role seems incorrect. 

1. There are growth opportunities in the current role that aren't about going to the next (Senior Engineer is a completely acceptable stopping point career wise, not everyone wants to go to Staff).
2. If growth is just about getting to the next level, we've not properly articulated the desired outcomes of growth.

Jordanjc3 commented on careers/technical-engineer.md:
```diff
@@ -1,73 +1,99 @@
+**Productivity:** A senior engineer regularly delivers a high volume of complex, critical work while also balancing other responsibilities or initiatives. The Senior Engineer requires very little oversight beyond high-level direction. They hold a depth of knowledge about the larger software ecosystem that enables them to debug quickly and effectively.
 
-The Senior Engineer has end-to-end responsibility for projects of increasing complexity that encompass more than their own development. They contribute to the common code bases and standards for the team. They understand the business that their code supports and use this knowledge to influence their task prioritization. Senior engineers assist in identifying and validating test cases and can iden ...
+**Impact:** Beyond just delivering work on our roadmap, a senior engineer helps implement innovations across our codebase, product, and department. They recognize and act upon opportunities that others miss. They seek to build objective and well-defended cases for decisions, and develop a reputation among the leadership team that their requests are objective, self-critical, and have the due-dilig ...
```
This is pretty good, just need to drop "They seek to build objective and well-defended....". The "outcome" here is either they do this or they don't, I'd expect a Senior to do this.

Jordanjc3 commented on careers/technical-engineer.md:
```diff
@@ -1,73 +1,99 @@
+Concrete examples could include:
 
-**Communication:** The Staff Engineer has strong abilities to influence without requiring reporting authority. They facilitate cross-team work and are influential beyond their individual group. They are capable of driving groups of disparate interests to decisions, and clearly communicating and seeing those decisions through to impact. The staff engineer is a very strong communicator and can simp ...
+- Having your perspective sought out in critical decisions
```
What's an example of a critical decision? Who's seeking out Staff Eng. perspective? VP? CTO? CEO?

Jordanjc3 commented on careers/technical-engineer.md:
```diff
@@ -1,73 +1,99 @@
 
-**Communication:** The Staff Engineer has strong abilities to influence without requiring reporting authority. They facilitate cross-team work and are influential beyond their individual group. They are capable of driving groups of disparate interests to decisions, and clearly communicating and seeing those decisions through to impact. The staff engineer is a very strong communicator and can simp ...
+- Having your perspective sought out in critical decisions
+- Actively inserting yourself into these critical moments
```
What does this mean in practice? Once we define critical moments, I'd like to understand what our expectations are around Staff Eng. "inserting [themselves]" into those critical moments.

Jordanjc3 commented on careers/technical-engineer.md:
```diff
@@ -1,73 +1,99 @@
+- decreasing onboarding time
+- getting the team to agree on coding standards
+- reviewing designs and architecture decisions
+- making us build the right thing
+- making our customers happier
```
These are not concrete, so I would make them so or drop them from the list of concrete examples.

Jordanjc3 commented on careers/technical-engineer.md:
```diff
@@ -1,73 +1,99 @@
-Packback's engineering ladder was inspired by documentation published by [Rent the Runway](https://dresscode.renttherunway.com/blog/ladder) and [Kickstarter](https://kickstarter.engineering/the-kickstarter-engineering-and-data-team-ladder-96996c3b327) on their engineering ladders. Thanks to them for open-sourcing that information.
+- Defining strategic technical initiatives
+- Solving technical problems of the highest scope and complexity for the entire organization
+- Actively seeking out and prioritize our toughest technical challenges with a goal of creating significant improvement
```
Instead of "with a goal of creating significant improvement", can we just say "and delivering significant improvement"

Jordanjc3 reviewed (changes requested):
Took my first pass. Overall this looks great! 

My biggest piece of feedback across the board is that we're using a lot of language like "has the ability to..." or "makes their best effort to...". For these competency models, our objective should be to create guidance that ensures both the Manager and their DR can clearly assess where they are across the levels by observing and pointing to visible behavior. Assessing intention or whether someone has the ability to do something is difficult, we can make this simpler by focusing on observable behaviors in our descriptions of the roles.

I left some feedback on the content within the levels and about adjusting the criteria, but my first pass here was more focused on _how_ we're writing these levels. I'll come back again once the adjustments have been made to dig into the substance of the expectations in each level.

Jordanjc3 commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+## Engineer II
 
-**Technical Scope:** Capable of taking well-scoped components from a larger project and completing these tasks in a reasonable time frame. Engineers at this level are assumed to be constantly making steady progress on tasks that are assigned to them and know when to ask for help when they are blocked. They can own their independent small-to-medium features all the way through from technical desig ...
+An engineer II is capable of taking on most technical problems and coordinating work spanning several sprints.
```
This is in conflict with the guidance laid out in the "Collaboration" section.

Jordanjc3 replied:
This says "coordinate work spanning multiple sprints", but "Impact" section states "are capable of independently planning work that multiple engineers can complete in a sprint or two"

dbhynds commented on careers/software-engineer.md:
```diff
@@ -4,52 +4,56 @@ The goal of the Software Engineer path is to train engineers to master the funda
+**Productivity:** They are capable of completing small, well-defined tasks with help from other engineers. Consulting with teammates before each ticket, they develop a plan to complete it on time. They can develop an initial plan to test their code but may need help completing it.
 
-**Communication:** The Apprentice is focused on learning the language of software development, and how to effectively communicate and work in an agile scrum environment. They should be a positive and curious presence on the team, working to improve themselves as engineers every day.
+**Impact:** They take on work that helps the team meet its sprint goals. They focus on learning and participating in the team's process, and consistently make their best attempt to follow the styles within each project.
```
At the apprentice level, I don't actually expect them to deliver work that aligns to our styles and guidelines. I expect them to still be learning these things. I expect them to try their best to do that tho.

Jordanjc3 replied:
For a given Engineering Manager having a performance discussion with an apprentice, how would they articulate whether that engineer is trying their best?

dbhynds commented on careers/managing-engineer.md:
```diff
@@ -82,7 +82,3 @@ The VP represents Engineering and Data on the senior team to articulate the capa
 * Develop cross-functional relationships with other business units (Sales, Retention, Revenue)
 * Standardize the development process where needed
 * Measure and improve the happiness of Engineering & ensure psychological safety
-
```
I plan to revisit this in the future, but not for this round of performance reviews.

dbhynds commented on careers/technical-engineer.md:
```diff
@@ -1,73 +1,99 @@
 ## Principal Engineer
 
-This person acts as the "chief architect" for our business. They have significant strategic vision and can take a high-level 3-5 year plan for growth at a business level and translate that into a strategic technology roadmap. They are deeply technically savvy, and their primary job is focusing on the architectural and technological needs to grow the business over the longer-term horizon. The prin ...
+This person acts as the "chief architect" for our business. They have significant strategic vision and can take a high-level 3-5 year plan for growth at a business level and translate that into a strategic technology roadmap. They are deeply technically savvy, and their primary job is focusing on the architectural and technological needs to grow the business over the longer-term horizon. The prin ...
```
Similar to the managing engineer path, the Principal Engineer title needs a revisit, but I want to do that another time. The changes made to this section were mostly to ensure consistent voice.

dbhynds commented:
For transparency sake:

I'm merging this PR even though I haven't addressed all the feedback. Reviews are imminently upon us, and I want to get this in front of the team ASAP.

There's still some very good feedback on this PR that I want to go back and address. That said, it's going to take more time and mental energy than I have capacity to do before reviews start. (They involve big re-thinks or re-writes.) I don't want to hold this up any longer.
